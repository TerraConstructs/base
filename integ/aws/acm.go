package aws

import (
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go/service/acm"
	"github.com/stretchr/testify/require"

	"github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/logger"
	"github.com/gruntwork-io/terratest/modules/retry"
	"github.com/gruntwork-io/terratest/modules/testing"
)

// ref: https://github.com/gruntwork-io/terratest/blob/v0.47.1/modules/aws/acm.go

// Get Certificate Status
func GetAcmCertificateStatus(t testing.TestingT, awsRegion string, certArn string) string {
	status, err := GetAcmCertificateStatusE(t, awsRegion, certArn)
	if err != nil {
		t.Fatal(err)
	}
	return status
}

// GetAcmCertificateStatusE gets the ACM certificate status for the given certificate ARN in the given region.
func GetAcmCertificateStatusE(t testing.TestingT, awsRegion string, certArn string) (string, error) {
	acmClient, err := aws.NewAcmClientE(t, awsRegion)
	if err != nil {
		return "", err
	}

	result, err := acmClient.DescribeCertificate(&acm.DescribeCertificateInput{
		CertificateArn: &certArn,
	})
	if err != nil {
		return "", err
	}

	return *result.Certificate.Status, nil
}

// WaitForCertificateIssued waits for the certificate to be issued
func WaitForCertificateIssued(
	t testing.TestingT,
	certArn string,
	region string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) {
	err := WaitForCertificateIssuedE(t, certArn, region, maxRetries, sleepBetweenRetries)
	require.NoError(t, err)
}

// WaitForCertificateIssuedE waits for the ACM Certificate to be issued
func WaitForCertificateIssuedE(
	t testing.TestingT,
	certArn string,
	region string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) error {
	msg, err := retry.DoWithRetryE(
		t,
		fmt.Sprintf("Waiting for Certificate %s to be ISSUED.", certArn),
		maxRetries,
		sleepBetweenRetries,
		func() (string, error) {
			certStatus, err := GetAcmCertificateStatusE(t, region, certArn)
			if err != nil {
				return "", err
			}
			if certStatus != acm.CertificateStatusIssued {
				return "", NewCertificateNotIssuedError(certArn, certStatus)
			}
			return fmt.Sprintf("Certificate %s is now at desired status %s", certArn, acm.CertificateStatusIssued), nil
		},
	)
	logger.Log(t, msg)
	return err
}
