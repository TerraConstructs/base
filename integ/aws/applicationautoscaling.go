package aws

import (
	"context"
	"fmt"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/applicationautoscaling"
	"github.com/aws/aws-sdk-go-v2/service/applicationautoscaling/types"
	"github.com/stretchr/testify/require"

	terratestaws "github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/testing"
)

// GetTableTrackingPolicy gets the target tracking policy for a DynamoDB table or errors if not found
func GetTableTrackingPolicy(t testing.TestingT, awsRegion string, resourceId string) *types.ScalingPolicy {
	policy, err := GetTableTrackingPolicyE(t, awsRegion, resourceId)
	require.NoError(t, err)
	return policy
}

// GetTableTrackingPolicy gets the target tracking policy for a DynamoDB table or returns an error if not found
func GetTableTrackingPolicyE(t testing.TestingT, awsRegion string, resourceId string) (*types.ScalingPolicy, error) {
	policies := GetScalingPolicies(t, awsRegion, "dynamodb")

	var targetTrackingPolicy *types.ScalingPolicy
	for _, policy := range policies {
		if *policy.ResourceId == resourceId &&
			policy.ScalableDimension == types.ScalableDimensionDynamoDBTableReadCapacityUnits &&
			policy.PolicyType == types.PolicyTypeTargetTrackingScaling {
			targetTrackingPolicy = &policy
			break
		}
	}
	if targetTrackingPolicy == nil {
		return nil, fmt.Errorf("no target tracking policy found for resource ID: %s", resourceId)
	}
	return targetTrackingPolicy, nil
}

// GetScalableTargets gets the Application Auto Scaling scalable targets for the given service namespace
func GetScalableTargets(t testing.TestingT, region string, serviceNamespace string) []types.ScalableTarget {
	targets, err := GetScalableTargetsE(t, region, serviceNamespace)
	require.NoError(t, err)
	return targets
}

// GetScalableTargetsE gets the Application Auto Scaling scalable targets for the given service namespace
func GetScalableTargetsE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScalableTarget, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalableTargets(context.Background(), &applicationautoscaling.DescribeScalableTargetsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScalableTargets, nil
}

// GetScalingPolicies gets the Application Auto Scaling scaling policies for the given service namespace
func GetScalingPolicies(t testing.TestingT, region string, serviceNamespace string) []types.ScalingPolicy {
	policies, err := GetScalingPoliciesE(t, region, serviceNamespace)
	require.NoError(t, err)
	return policies
}

// GetScalingPoliciesE gets the Application Auto Scaling scaling policies for the given service namespace
func GetScalingPoliciesE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScalingPolicy, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalingPolicies(context.Background(), &applicationautoscaling.DescribeScalingPoliciesInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScalingPolicies, nil
}

// GetScheduledActions gets the Application Auto Scaling scheduled actions for the given service namespace
func GetScheduledActions(t testing.TestingT, region string, serviceNamespace string) []types.ScheduledAction {
	actions, err := GetScheduledActionsE(t, region, serviceNamespace)
	require.NoError(t, err)
	return actions
}

// GetScheduledActionsE gets the Application Auto Scaling scheduled actions for the given service namespace
func GetScheduledActionsE(t testing.TestingT, region string, serviceNamespace string) ([]types.ScheduledAction, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScheduledActions(context.Background(), &applicationautoscaling.DescribeScheduledActionsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
	})

	if err != nil {
		return nil, err
	}

	return result.ScheduledActions, nil
}

// GetScalableTargetsByResourceId gets scalable targets filtered by resource ID
func GetScalableTargetsByResourceId(t testing.TestingT, region string, serviceNamespace string, resourceId string) []types.ScalableTarget {
	targets, err := GetScalableTargetsByResourceIdE(t, region, serviceNamespace, resourceId)
	require.NoError(t, err)
	return targets
}

// GetScalableTargetsByResourceIdE gets scalable targets filtered by resource ID
func GetScalableTargetsByResourceIdE(t testing.TestingT, region string, serviceNamespace string, resourceId string) ([]types.ScalableTarget, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScalableTargets(context.Background(), &applicationautoscaling.DescribeScalableTargetsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
		ResourceIds:      []string{resourceId},
	})

	if err != nil {
		return nil, err
	}

	return result.ScalableTargets, nil
}

// GetScheduledActionsByResourceId gets scheduled actions filtered by resource ID
func GetScheduledActionsByResourceId(t testing.TestingT, region string, serviceNamespace string, resourceId string) []types.ScheduledAction {
	actions, err := GetScheduledActionsByResourceIdE(t, region, serviceNamespace, resourceId)
	require.NoError(t, err)
	return actions
}

// GetScheduledActionsByResourceIdE gets scheduled actions filtered by resource ID
func GetScheduledActionsByResourceIdE(t testing.TestingT, region string, serviceNamespace string, resourceId string) ([]types.ScheduledAction, error) {
	client, err := NewApplicationAutoScalingClientE(t, region)
	if err != nil {
		return nil, err
	}

	result, err := client.DescribeScheduledActions(context.Background(), &applicationautoscaling.DescribeScheduledActionsInput{
		ServiceNamespace: types.ServiceNamespace(serviceNamespace),
		ResourceId:       aws.String(resourceId),
	})

	if err != nil {
		return nil, err
	}

	return result.ScheduledActions, nil
}

// NewApplicationAutoScalingClient creates a new Application Auto Scaling client
func NewApplicationAutoScalingClient(t testing.TestingT, region string) *applicationautoscaling.Client {
	client, err := NewApplicationAutoScalingClientE(t, region)
	require.NoError(t, err)
	return client
}

// NewApplicationAutoScalingClientE creates a new Application Auto Scaling client
func NewApplicationAutoScalingClientE(t testing.TestingT, region string) (*applicationautoscaling.Client, error) {
	sess, err := terratestaws.NewAuthenticatedSession(region)
	if err != nil {
		return nil, err
	}
	return applicationautoscaling.NewFromConfig(*sess), nil
}
