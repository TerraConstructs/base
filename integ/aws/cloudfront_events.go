package aws

import (
	"encoding/json"
	"fmt"
	"os"
)

type Version string

const Version1_0 Version = "1.0"

func (v *Version) UnmarshalJSON(data []byte) error {
	var versionStr string
	if err := json.Unmarshal(data, &versionStr); err != nil {
		return err
	}
	if versionStr != string(Version1_0) {
		return fmt.Errorf("unsupported version: %s", versionStr)
	}
	*v = Version(versionStr)
	return nil
}

func (v Version) MarshalJSON() ([]byte, error) {
	if v != Version1_0 {
		return nil, fmt.Errorf("unsupported version: %s", v)
	}
	return json.Marshal(string(v))
}

// CloudFrontFunctionEvent represents the event object that you provide to test a CloudFront function.
// https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/functions-event-structure.html
type CloudFrontFunctionEvent struct {
	Version  Version   `json:"version"` // The version field contains a string that specifies the version of the CloudFront Functions event object. The current version is 1.0.
	Context  Context   `json:"context"` // The context object contains contextual information about the event
	Viewer   Viewer    `json:"viewer"`
	Request  *Request  `json:"request,omitempty"`
	Response *Response `json:"response,omitempty"`
}

type Context struct {
	DistributionDomainName string `json:"distributionDomainName"` // The CloudFront domain name (for example, d111111abcdef8.cloudfront.net) of the distribution that's associated with the event.
	DistributionID         string `json:"distributionId"`
	EventType              string `json:"eventType"`
	RequestID              string `json:"requestId"`
}

type Viewer struct {
	IP string `json:"ip"`
}

type ValueObject map[string]ValueEntry

type ValueEntry struct {
	Value      string       `json:"value"`
	MultiValue []MultiValue `json:"multiValue,omitempty"`
}

type MultiValue struct {
	Value string `json:"value"`
}

type Request struct {
	Method      string      `json:"method"`
	URI         string      `json:"uri"`
	Querystring ValueObject `json:"querystring"`
	Headers     ValueObject `json:"headers"`
	Cookies     ValueObject `json:"cookies,omitempty"`
}

type Response struct {
	StatusCode        int             `json:"statusCode"`
	StatusDescription *string         `json:"statusDescription,omitempty"`
	Headers           *ValueObject    `json:"headers,omitempty"`
	Cookies           *ResponseCookie `json:"cookies,omitempty"`
}

type ResponseCookie map[string]ResponseCookieEntry

type ResponseCookieEntry struct {
	Value      string            `json:"value"`
	Attributes string            `json:"attributes"`
	MultiValue []MultiValueAttrs `json:"multiValue,omitempty"`
}

type MultiValueAttrs struct {
	Value      string `json:"value"`
	Attributes string `json:"attributes"`
}

func ReadCloudFrontEvent(path string) (*CloudFrontFunctionEvent, error) {
	f, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var t = CloudFrontFunctionEvent{
		Version: "1.0",
		Context: Context{
			EventType: "viewer-request",
		},
		Viewer: Viewer{
			IP: "1.2.3.4",
		},
	}

	err = json.Unmarshal(f, &t)
	return &t, err
}
