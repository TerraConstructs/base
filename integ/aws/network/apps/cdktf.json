{
  "//": "Dummy file to satisfy the bundling findUp logic"
}
