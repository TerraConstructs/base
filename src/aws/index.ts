// core
// export * from "./tag-aspect";
// export * from "./tag-manager";
export * from "./arn";
export * from "./aws-construct";
export * from "./aws-stack";
export * from "./aws-tags";
export * from "./log-retention";

// core JSII compatible generated interfaces
export * from "./provider-config.generated";

// classes
export * as iam from "./iam";
export * as encryption from "./encryption";
export * as compute from "./compute";
export * as network from "./network";
export * as storage from "./storage";
export * as notify from "./notify";
export * as edge from "./edge";
export * as cloudwatch from "./cloudwatch";
