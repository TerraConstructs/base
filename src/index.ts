export * from "./construct-base";
export * from "./stack-base";
export * from "./duration";
export * from "./size";

// cloud specific exports
export * as aws from "./aws";
