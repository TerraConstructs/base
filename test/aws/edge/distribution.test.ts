import {
  s3BucketWebsiteConfiguration,
  dataAwsIamPolicyDocument,
} from "@cdktf/provider-aws";
import { App, Testing } from "cdktf";
import "cdktf/lib/testing/adapters/jest";
import { edge, storage, AwsStack } from "../../../src/aws";
import { Template } from "../../assertions";

const environmentName = "Test";
const gridUUID = "123e4567-e89b-12d3";
const gridBackendConfig = {
  address: "http://localhost:3000",
};
const providerConfig = { region: "us-east-1" };
describe("Distribution", () => {
  let app: App;
  let stack: AwsStack;

  beforeEach(() => {
    app = Testing.app();
    stack = new AwsStack(app, "MyStack", {
      environmentName,
      gridUUID,
      providerConfig,
      gridBackendConfig,
      // TODO: Should support passing account via Stack props to match AWS CDK cross account support
      // account: "1234",
    });
  });

  test("Should synth with OAI and match SnapShot", () => {
    // GIVEN
    const bucket = new storage.Bucket(stack, "HelloWorld", {
      namePrefix: "hello-world",
      cloudfrontAccess: {
        enabled: true,
      },
    });
    // WHEN
    new edge.Distribution(stack, "HelloWorldDistribution", {
      defaultBehavior: {
        origin: new edge.S3Origin(bucket),
      },
    });
    // THEN
    Template.synth(stack, { snapshot: true }).toHaveDataSourceWithProperties(
      dataAwsIamPolicyDocument.DataAwsIamPolicyDocument,
      {
        statement: [
          {
            actions: ["s3:GetObject"],
            effect: "Allow",
            principals: [
              {
                identifiers: [
                  "${aws_cloudfront_origin_access_identity.HelloWorld_OriginAccessIdentity_5B20D425.iam_arn}",
                ],
                type: "AWS",
              },
            ],
            resources: [`${stack.resolve(bucket.bucketArn)}/*`],
          },
        ],
      },
    );
  });
  test("Should synth with websiteConfig and match SnapShot", () => {
    // GIVEN
    const bucket = new storage.Bucket(stack, "HelloWorld", {
      namePrefix: "hello-world",
      websiteConfig: {
        enabled: true,
      },
    });
    // WHEN
    new edge.Distribution(stack, "HelloWorldDistribution", {
      defaultBehavior: {
        origin: new edge.S3Origin(bucket),
      },
    });
    // THEN
    Template.synth(stack, { snapshot: true }).toHaveResourceWithProperties(
      s3BucketWebsiteConfiguration.S3BucketWebsiteConfiguration,
      {
        bucket: stack.resolve(bucket.bucketName),
      },
    );
  });
  test("Should throw error if bucket has no OAI or website config", () => {
    // WHEN
    const bucket = new storage.Bucket(stack, "HelloWorld", {
      namePrefix: "hello-world",
    });
    // THEN
    expect(() => {
      new edge.Distribution(stack, "HelloWorldDistribution", {
        defaultBehavior: {
          origin: new edge.S3Origin(bucket),
        },
      });
    }).toThrow("must have an origin access identity");
  });
  test("Should support multiple origins and cache behaviors", () => {
    // GIVEN
    const bucket0 = new storage.Bucket(stack, "Bucket0", {
      namePrefix: "bucket-0",
      websiteConfig: {
        enabled: true,
      },
    });
    const bucket1 = new storage.Bucket(stack, "Bucket1", {
      namePrefix: "bucket-1",
      websiteConfig: {
        enabled: true,
      },
    });
    // WHEN
    new edge.Distribution(stack, "HelloWorldDistribution", {
      defaultBehavior: {
        origin: new edge.S3Origin(bucket0),
      },
      additionalBehaviors: {
        "/images/*": {
          origin: new edge.S3Origin(bucket1),
        },
      },
    });
    // THEN
    Template.fromStack(stack, { snapshot: true }).toMatchObject({
      resource: {
        aws_s3_bucket_website_configuration: {
          Bucket0_WebsiteConfig_F3339C3F: {
            bucket: stack.resolve(bucket0.bucketName),
            index_document: {
              suffix: "index.html",
            },
          },
          Bucket1_WebsiteConfig_0DE2B7DD: {
            bucket: stack.resolve(bucket1.bucketName),
            index_document: {
              suffix: "index.html",
            },
          },
        },
      },
    });
  });
  test("Should support custom Response Header Policy", () => {
    // GIVEN
    const bucket = new storage.Bucket(stack, "Bucket", {
      namePrefix: "bucket",
      cloudfrontAccess: {
        enabled: true,
      },
    });
    // With COOP/COEP headers
    const responseHeadersPolicy = new edge.ResponseHeadersPolicy(
      stack,
      "ResponseHeadersPolicy",
      {
        responseHeadersPolicyName: "CrossOriginIsolation",
        // ref: https://webcontainers.io/guides/configuring-headers
        customHeadersBehavior: {
          customHeaders: [
            {
              header: "Cross-Origin-Embedder-Policy",
              value: "require-corp",
              override: true,
            },
            {
              header: "Cross-Origin-Opener-Policy",
              value: "same-origin",
              override: true,
            },
          ],
        },
      },
    );
    // WHEN
    new edge.Distribution(stack, "HelloWorldDistribution", {
      defaultBehavior: {
        origin: new edge.S3Origin(bucket),
        responseHeadersPolicy,
      },
    });
    // THEN
    Template.fromStack(stack).toMatchObject({
      resource: {
        aws_cloudfront_distribution: {
          HelloWorldDistribution_E7735130: {
            default_cache_behavior: {
              response_headers_policy_id: stack.resolve(
                responseHeadersPolicy.responseHeadersPolicyId,
              ),
            },
          },
        },
      },
    });
  });
});
