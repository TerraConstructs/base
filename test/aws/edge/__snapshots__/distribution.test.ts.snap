// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Distribution Should support multiple origins and cache behaviors 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_cloudfront_cache_policy": {
      "HelloWorldDistribution_Managed-CachingOptimized_E039BAC9": {
        "name": "Managed-CachingOptimized"
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "resource": {
    "aws_cloudfront_distribution": {
      "HelloWorldDistribution_E7735130": {
        "default_cache_behavior": {
          "allowed_methods": [
            "GET",
            "HEAD"
          ],
          "cache_policy_id": "\${data.aws_cloudfront_cache_policy.HelloWorldDistribution_Managed-CachingOptimized_E039BAC9.id}",
          "cached_methods": [
            "GET",
            "HEAD"
          ],
          "compress": true,
          "target_origin_id": "origin-0",
          "viewer_protocol_policy": "allow-all"
        },
        "enabled": true,
        "http_version": "http2",
        "ordered_cache_behavior": [
          {
            "allowed_methods": [
              "GET",
              "HEAD"
            ],
            "cache_policy_id": "\${data.aws_cloudfront_cache_policy.HelloWorldDistribution_Managed-CachingOptimized_E039BAC9.id}",
            "cached_methods": [
              "GET",
              "HEAD"
            ],
            "compress": true,
            "path_pattern": "/images/*",
            "target_origin_id": "origin-1",
            "viewer_protocol_policy": "allow-all"
          }
        ],
        "origin": [
          {
            "custom_origin_config": {
              "http_port": 80,
              "https_port": 443,
              "origin_protocol_policy": "http-only",
              "origin_ssl_protocols": [
                "TLSv1.2"
              ]
            },
            "domain_name": "\${aws_s3_bucket_website_configuration.Bucket0_WebsiteConfig_F3339C3F.website_endpoint}",
            "origin_id": "origin-0"
          },
          {
            "custom_origin_config": {
              "http_port": 80,
              "https_port": 443,
              "origin_protocol_policy": "http-only",
              "origin_ssl_protocols": [
                "TLSv1.2"
              ]
            },
            "domain_name": "\${aws_s3_bucket_website_configuration.Bucket1_WebsiteConfig_0DE2B7DD.website_endpoint}",
            "origin_id": "origin-1"
          }
        ],
        "restrictions": {
          "geo_restriction": {
            "restriction_type": "none"
          }
        },
        "tags": {
          "Name": "Test-HelloWorldDistribution",
          "grid:EnvironmentName": "Test",
          "grid:UUID": "123e4567-e89b-12d3"
        },
        "viewer_certificate": {
          "cloudfront_default_certificate": true,
          "minimum_protocol_version": "TLSv1.2_2021",
          "ssl_support_method": "sni-only"
        }
      }
    },
    "aws_s3_bucket": {
      "Bucket0_C98EAF71": {
        "bucket_prefix": "bucket-0-mystackbucket0",
        "tags": {
          "Name": "Test-Bucket0",
          "grid:EnvironmentName": "Test",
          "grid:UUID": "123e4567-e89b-12d3"
        }
      },
      "Bucket1_2520700A": {
        "bucket_prefix": "bucket-1-mystackbucket1",
        "tags": {
          "Name": "Test-Bucket1",
          "grid:EnvironmentName": "Test",
          "grid:UUID": "123e4567-e89b-12d3"
        }
      }
    },
    "aws_s3_bucket_website_configuration": {
      "Bucket0_WebsiteConfig_F3339C3F": {
        "bucket": "\${aws_s3_bucket.Bucket0_C98EAF71.bucket}",
        "index_document": {
          "suffix": "index.html"
        }
      },
      "Bucket1_WebsiteConfig_0DE2B7DD": {
        "bucket": "\${aws_s3_bucket.Bucket1_2520700A.bucket}",
        "index_document": {
          "suffix": "index.html"
        }
      }
    }
  },
  "terraform": {
    "backend": {
      "http": {
        "address": "http://localhost:3000"
      }
    },
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.88.0"
      }
    }
  }
}"
`;

exports[`Distribution Should synth with OAI and match SnapShot 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_cloudfront_cache_policy": {
      "HelloWorldDistribution_Managed-CachingOptimized_E039BAC9": {
        "name": "Managed-CachingOptimized"
      }
    },
    "aws_iam_policy_document": {
      "HelloWorld_Policy_9607E751": {
        "statement": [
          {
            "actions": [
              "s3:GetObject"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "\${aws_cloudfront_origin_access_identity.HelloWorld_OriginAccessIdentity_5B20D425.iam_arn}"
                ],
                "type": "AWS"
              }
            ],
            "resources": [
              "\${aws_s3_bucket.HelloWorld_7964D1E8.arn}/*"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "resource": {
    "aws_cloudfront_distribution": {
      "HelloWorldDistribution_E7735130": {
        "default_cache_behavior": {
          "allowed_methods": [
            "GET",
            "HEAD"
          ],
          "cache_policy_id": "\${data.aws_cloudfront_cache_policy.HelloWorldDistribution_Managed-CachingOptimized_E039BAC9.id}",
          "cached_methods": [
            "GET",
            "HEAD"
          ],
          "compress": true,
          "target_origin_id": "origin-0",
          "viewer_protocol_policy": "allow-all"
        },
        "enabled": true,
        "http_version": "http2",
        "origin": [
          {
            "domain_name": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket_regional_domain_name}",
            "origin_id": "origin-0",
            "s3_origin_config": {
              "origin_access_identity": "\${aws_cloudfront_origin_access_identity.HelloWorld_OriginAccessIdentity_5B20D425.cloudfront_access_identity_path}"
            }
          }
        ],
        "restrictions": {
          "geo_restriction": {
            "restriction_type": "none"
          }
        },
        "tags": {
          "Name": "Test-HelloWorldDistribution",
          "grid:EnvironmentName": "Test",
          "grid:UUID": "123e4567-e89b-12d3"
        },
        "viewer_certificate": {
          "cloudfront_default_certificate": true,
          "minimum_protocol_version": "TLSv1.2_2021",
          "ssl_support_method": "sni-only"
        }
      }
    },
    "aws_cloudfront_origin_access_identity": {
      "HelloWorld_OriginAccessIdentity_5B20D425": {
        "comment": "OAI for \${aws_s3_bucket.HelloWorld_7964D1E8.bucket}"
      }
    },
    "aws_s3_bucket": {
      "HelloWorld_7964D1E8": {
        "bucket_prefix": "hello-world-mystackhelloworld",
        "tags": {
          "Name": "Test-HelloWorld",
          "grid:EnvironmentName": "Test",
          "grid:UUID": "123e4567-e89b-12d3"
        }
      }
    },
    "aws_s3_bucket_policy": {
      "HelloWorld_Policy_9E56801E": {
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "policy": "\${data.aws_iam_policy_document.HelloWorld_Policy_9607E751.json}"
      }
    }
  },
  "terraform": {
    "backend": {
      "http": {
        "address": "http://localhost:3000"
      }
    },
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.88.0"
      }
    }
  }
}"
`;

exports[`Distribution Should synth with websiteConfig and match SnapShot 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_cloudfront_cache_policy": {
      "HelloWorldDistribution_Managed-CachingOptimized_E039BAC9": {
        "name": "Managed-CachingOptimized"
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "resource": {
    "aws_cloudfront_distribution": {
      "HelloWorldDistribution_E7735130": {
        "default_cache_behavior": {
          "allowed_methods": [
            "GET",
            "HEAD"
          ],
          "cache_policy_id": "\${data.aws_cloudfront_cache_policy.HelloWorldDistribution_Managed-CachingOptimized_E039BAC9.id}",
          "cached_methods": [
            "GET",
            "HEAD"
          ],
          "compress": true,
          "target_origin_id": "origin-0",
          "viewer_protocol_policy": "allow-all"
        },
        "enabled": true,
        "http_version": "http2",
        "origin": [
          {
            "custom_origin_config": {
              "http_port": 80,
              "https_port": 443,
              "origin_protocol_policy": "http-only",
              "origin_ssl_protocols": [
                "TLSv1.2"
              ]
            },
            "domain_name": "\${aws_s3_bucket_website_configuration.HelloWorld_WebsiteConfig_72ADC681.website_endpoint}",
            "origin_id": "origin-0"
          }
        ],
        "restrictions": {
          "geo_restriction": {
            "restriction_type": "none"
          }
        },
        "tags": {
          "Name": "Test-HelloWorldDistribution",
          "grid:EnvironmentName": "Test",
          "grid:UUID": "123e4567-e89b-12d3"
        },
        "viewer_certificate": {
          "cloudfront_default_certificate": true,
          "minimum_protocol_version": "TLSv1.2_2021",
          "ssl_support_method": "sni-only"
        }
      }
    },
    "aws_s3_bucket": {
      "HelloWorld_7964D1E8": {
        "bucket_prefix": "hello-world-mystackhelloworld",
        "tags": {
          "Name": "Test-HelloWorld",
          "grid:EnvironmentName": "Test",
          "grid:UUID": "123e4567-e89b-12d3"
        }
      }
    },
    "aws_s3_bucket_website_configuration": {
      "HelloWorld_WebsiteConfig_72ADC681": {
        "bucket": "\${aws_s3_bucket.HelloWorld_7964D1E8.bucket}",
        "index_document": {
          "suffix": "index.html"
        }
      }
    }
  },
  "terraform": {
    "backend": {
      "http": {
        "address": "http://localhost:3000"
      }
    },
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.88.0"
      }
    }
  }
}"
`;
